lib/src/gc.rs

```diff
@@ -98,106 +91,129 @@ impl<'r, I, J> CollectableRefs<'r, I, J>
-                // here, in the exact same way it is for leaves. We do that
-                // because can't mix the computation with those of the leaves.
-                // It would cause head references to be removed if any message
-                // was posted as a reply to the current head.
-                let mut head_history = self
-                    .repo
-                    .revwalk()
-                    .chain_err(|| EK::CannotConstructRevwalk)?;
-                match self.collect_heads {
-                    ReferenceCollectionSpec::Never => {},
-                    ReferenceCollectionSpec::BackedByRemoteHead => {
-                        for item in issue.remote_refs(IssueRefType::Head)? {
-                            head_history.push(
-                                item?
-                                    .peel(git2::ObjectType::Commit)
-                                    .chain_err(|| EK::CannotGetCommit)?
-                                    .id()
-                            )?;
-                        }
-                    },
-                };
-                let mut referring_refs = iter::RefsReferringTo::new(head_history);
-                referring_refs.watch_ref(local_head)?;
-                referring_refs.collect_result_into(&mut retval)?;
+        // remote refs
+        if self.consider_remote_refs {
+            for item in issue.remote_refs(IssueRefType::Any)? {
+                retval.push(item?
+                    .peel(git2::ObjectType::Commit)
+                    .chain_err(|| EK::CannotGetCommit)?
+                    .id()
+                )?;
             }
+        }
 
-            // local leaves
-            for item in issue.local_refs(IssueRefType::Leaf)? {
-                let leaf = item?;
-                // NOTE: We push the parents of the references rather than the
-                //       references themselves since that would cause the
-                //       `RefsReferringTo` report that exact same reference.
-                Self::push_ref_parents(&mut messages, &leaf)?;
-                refs_to_assess.push(leaf);
-            }
+        Ok(retval)
+    }
 
-            // remote refs
-            if self.consider_remote_refs {
-                for item in issue.remote_refs(IssueRefType::Any)? {
-                    messages.push(item?
-                        .peel(git2::ObjectType::Commit)
-                        .chain_err(|| EK::CannotGetCommit)?
-                        .id()
-                    )?;
-                }
-            }
-        }
+    /// Find collectable references for multiple issues
+    ///
+    /// This is a convenience function.
+    ///
+    /// # Note
+    ///
+    /// Internally, this function collects the references during the call.
+    /// Consider using the `for_issues` function in conjunction with
+    /// `flat_map()` on an issue iterator instead.
+    ///
+    pub fn into_refs<I, J, K>(self, issues: I) -> Result<Vec<Reference<'r>>>
```

Comment:
I don't understand why this does return a `Vec<_>` instead of an custom iterator type.

If this would return a custom iterator type, the `into_collector()` function would also return a custom iterator - which is what it does, and the `ReferenceCollector` type stores an iterator internally. So effectively, we create a `Vec` that we turn into an iterator type again in case of the `into_collector()` function.

In case of the other use of this function, which is in `push_ref_parents()`, we turn the `Vec` directly into an iterator again and do some computation on it, just to collect it afterwards. Which is not very efficient either.

So I guess it would be really beneficial to let this function return an iterator type as well!

Reply:
The plan is to remove this function, eventually. So I don't see any point in adding yet another custom iterator just for this function.

Reply:
Mid-May the impl-trait feature will be in stable Rust. We can then move forward and return a trait here, I guess.

Reply:
Mh... I'll probably re-iterate over the function, and maybe remove the deprecated-annotation I just applied.

We have a few other places where we would greatly benefit from an impl-trait feature btw.